add functionality to convert windows paths (all shapes and sizes) to `loc` values

jurgenvinju opened the pull request:
Rethinking this based on feedback from @DavyLandman ; the current PR splits the responsibility of interpreting Windows file paths into a syntactical and a semantic part. Also it supports all shapes and sizes, including UNC paths, and does not normalize automatically.

* syntax definition in lang::paths::Windows reverse engineered from docs and open-source path parsers in Java and C#; this is the syntactic part. 
* simple syntactic convertor function in lang::paths::Windows deals with the 5 kinds of paths, mapping them to `file:///`, `cwd:///`, `cwdrive:///` and `unc:///` paths
   * warns about illegal paths via a parse error (e.g. trailing "." or whitespace)
   * interprets different prefix notations `C:`, `C:\`, `\`, `\\` mapping them to different schemes
   * skips all additional slashes or backslashes, as it should
   * does no decoding of the path literals, but everything is _encoded_ for URI syntax automatically. E.g. `file:///C:/Program%20Files`  is the result of converting `C:\Program Files\`
 * For the semantic interpretation (actual file IO), we've had to add two new resolvers to the team:
    * `|cwdrive:///|` refers to the root of the file system of the current working directory. On Windows that would be `D:/` or `C:/` for example, and on Unix it is always `/`. Not useful for unix, but essential to be able to encode the `\filename` notations of Windows paths.
    * `|unc://<machine>/<share>/<path>|` is mapped from `\\<machine>\<share>\<path>` and back, such that UNC paths actually work on Windows systems. This _only works_ on Windows systems. We could have used the empty authority position of `|file:///|`, however that provides a false sense of portability that we do not have. `|unc:///|` literals work  as identifiers on any system, but only for file IO on Windows systems. 
       * Note that `|unc:///|` paths have the annoying property that machine names are valid only for the local network, so they are just as machine-specific as the `file:///` scheme is. 
       * The actual implementation derives directly from the `file:///` scheme because `new File` and friends from the Java SDK implement the UNC scheme natively.
 
 Added some tests to this effect, but have to run them on Windows.

DavyLandman commented on src/org/rascalmpl/library/Prelude.java:
```diff
@@ -3615,6 +3618,70 @@ public ISet findResources(IString fileName) {
+
+		// generate either a root file loc or a current working directory loc.
+		if (path.startsWith(sep) && drive.isEmpty()) {
+			return URIUtil.correctLocation("file", "", drive + path);
```
shouldn't this be `drive + ":/" + path`?

jurgenvinju replied:
No. The colon is included.

DavyLandman commented on src/org/rascalmpl/library/Prelude.java:
```diff
@@ -3615,6 +3618,70 @@ public ISet findResources(IString fileName) {
+			return URIUtil.correctLocation("file", "", drive + path);
+		}
+		else {
+			return URIUtil.correctLocation("cwd", "", drive + path);
```
`cwd` and non empty drive doesn't seem to make sense?

jurgenvinju replied:
Indeed. That's up to a windows feature that `C:name` means C:\currentDirOnC\name`

jurgenvinju commented on src/org/rascalmpl/library/Location.rsc:
```diff
@@ -43,6 +43,63 @@ loc relativize(list[loc] haystack, loc needle) {
+
+* if `legalize` is true, then every path segment is legalized by replacing offending characters to `_`
+* if `legalize` is false, and offending characters appear between the path separators, then an IO exception is thrown.
+* if the requested file system syntax is either case insensitive or not case preserving (or both), then all uppercase characters will be replaced by lowercase characters.
```
We have to think about this syntax without an actual filesystem to test properties on. That's why this is hard.

I can turn it off, or make it configurable but only if the filesystem type supports it, or even allow the use to force normalized cases. In any case it will always be a problematic feature; without normalization can also be very bad the accuracy of downstream analysis on locs... 

Note that we do not normalize on windows because that would break the link with the actual filesystem.

jurgenvinju commented:
> I think this function is not ready. Especially for Windows the cases are not thought out.
> 
This is everything we can leach out of Apache Commons. I don't see the sense in building more of our own string manipulations on top. Rather let's find another library that implements what we need. At least the OS specific parts `new File` do exactly what we need, but those algorithms aren't accessible from other OSes.. we could clone them?

> We need at least some test that show how the library is supposed to work.

There is already one example in the docs; but can only uncomment after a bootstap.

jurgenvinju commented:
That library search turned up nothing. Another option I'm considering is to source the standards for each path syntax and write a small syntax definition for each available format. That way at least it's fully transparent what we can process. It's more work, but also it's "the Rascal way", so serves as an example in the library.

jurgenvinju commented on src/org/rascalmpl/library/Prelude.java:
```diff
@@ -3615,6 +3618,70 @@ public ISet findResources(IString fileName) {
+		}
+
+		// generate either a root file loc or a current working directory loc.
+		if (path.startsWith(sep) && drive.isEmpty()) {
```
Drive is only non empty for windows. When the path starts with / and it's not windows then it's an absolute path.
In windows relative paths that don't start with / are encoded differently, not with cwd but as `file:///C:name` because every drive can have it's own current working directory.

jurgenvinju replied:
I think we'd better add a`cwd` optional parameter to indicate which cwd is to be used to resolve relative paths in. 

Keeping in mind that we can never test against actual filesystems for this function.

jurgenvinju commented on src/org/rascalmpl/library/Location.rsc:
```diff
@@ -43,6 +43,63 @@ loc relativize(list[loc] haystack, loc needle) {
+* if `legalize` is false, and offending characters appear between the path separators, then an IO exception is thrown.
+* if the requested file system syntax is either case insensitive or not case preserving (or both), then all uppercase characters will be replaced by lowercase characters.
+* on windows systems the drive letter `C:` is added if a drive letter is missing.
+* on all the other systems, if the path starts with a path separator, it is taken as absolute in `file:///`. Otherwise the root will be `cwd:///`.
```
Yes we could also support that, however this library from Apache does not and UNC has many particularities that make it quite a different story to cover to URI. I believe we should reuse a library for that or reuse java `new File` or `new URL` that does support UNC. However,  these two are extremely slow because they insist on setting up a network connection. All we want here is a parser and an isomorphic translation... 

The jury is still out. But I'm leaning towards a full syntax definition with a concrete tree processor for the mapping.

DavyLandman commented:
> There is already one example in the docs; but can only uncomment after a bootstap.

We don't need to Bootstrap to have unit tests right? We should be able to write tests that say, given this path and this file system, we expect this loc? And then a bunch of examples for it?

jurgenvinju commented:
> > There is already one example in the docs; but can only uncomment after a bootstap.
> 
> We don't need to Bootstrap to have unit tests right? We should be able to write tests that say, given this path and this file system, we expect this loc? And then a bunch of examples for it?

Yes sure; there are many tests to be added. I forgot to say that i agreed. However with the current feedback I'm not sure that leaning on Apache Commons is good enough; and there is no implementation independent spec of what to test for yet. So I'm going to search for that and elevate the spec to tests and a syntax definition.


There is"pathlibl in the stdlib of python that comes very close. Will study that for a bit.

jurgenvinju commented:
Found the exact code that we are looking for here: https://github.com/frohoff/jdk8u-jdk/blob/master/src/windows/classes/sun/nio/fs/WindowsPathParser.java

That's GPL licensed. So these three algorithms (parser and normalize and finally toURI) should be mapped exactly to Rascal to get a very reasonable mapping. That should also guarantee that access of these files via the `file:///` scheme works as expected. If that can not be achieved then different protocols should be introduced, like `unc:///` that can interpret the encoding back to where it came from. I'm hoping for now the `file` scheme will do, however.

jurgenvinju commented:
Rewrote from scratch; worked all the feedback from @DavyLandman into it, I believe. Also I'm focusing only on Windows paths from now on. With the current design we can add support for other Path syntaxes incrementally and in different modules.

jurgenvinju commented:
Note bene: I wrote the tests in the library module that provides the functionality; however I'm pretty sure these are not triggered by the current CI setup.

DavyLandman commented on src/org/rascalmpl/library/Location.rsc:
```diff
@@ -43,6 +44,18 @@ loc relativize(list[loc] haystack, loc needle) {
+* Relative drive-specific: `C:hello.txt`
+* Relative: `hello.txt`
+* Directory-relative: `\hello.txt`
+* UNC format: `\\\\system07\\C$\\`
```
the other 2 lines (53 & 50) don't escape the `\`?

jurgenvinju replied:
check! copy/paste issue. I'd like the literal value in the docs even though the string version has to be escaped.

DavyLandman commented on src/org/rascalmpl/library/Location.rsc:
```diff
@@ -43,6 +44,18 @@ loc relativize(list[loc] haystack, loc needle) {
+* Directory-relative: `\hello.txt`
+* UNC format: `\\\\system07\\C$\\`
+
+Windows paths, againat popular believe, support both `/` and `\` as path separators.
```
`s/againat/against/`

apart from that, yes, it has always been supported. and you can even mix them..

DavyLandman commented on src/org/rascalmpl/library/Prelude.java:
```diff
@@ -3615,7 +3615,15 @@ public ISet findResources(IString fileName) {
 	}
-	
+
+	public IConstructor currentFileSystem() {
```
do we still need this function here?

jurgenvinju replied:
no. thanks!

DavyLandman commented on src/org/rascalmpl/library/lang/paths/Windows.rsc:
```diff
@@ -0,0 +1,113 @@
+This is left to downstream processors of `loc` values, if necessary. The current transformation
+is purely syntactical, and tries to preserve the semantics of the path as much as possible.
+}
+module lang::paths::Windows
```
![](https://media1.tenor.com/m/WB0asE2__fAAAAAd/the-book-of-boba-fett-this-is-the-way.gif)

jurgenvinju replied:
hahaha. yes it felt like this. a homecoming.

DavyLandman commented on src/org/rascalmpl/library/lang/paths/Windows.rsc:
```diff
@@ -0,0 +1,113 @@
+loc parseWindowsPath(str input, loc src=|unknown:///|) = mapPathToLoc(parse(#WindowsPath, input, src));
+
+@synopsis{UNC}
+loc mapPathToLoc((WindowsPath) `<Slash _><Slash _><Slashes? _><PathChar* hostName><Slashes _><PathChar* shareName><Slashes _><WindowsFilePath path>`)
```
Do we want to normalize that if hostname is `?` or `.` that it's just a `file:///` uri?

jurgenvinju replied:
I thought about that but no that's not possible in general because the ? notation has (very) different length limitations, and I don't want a length-dependent normalization process. To finicky.

DavyLandman replied:
ah true.

DavyLandman commented on src/org/rascalmpl/library/lang/paths/Windows.rsc:
```diff
@@ -0,0 +1,113 @@
+
+@synopsis{Absolute: given the drive and relative to its root.}
+loc mapPathToLoc((WindowsPath) `<Drive drive>:<Slashes _><WindowsFilePath path>`) 
+    = (|file:///<drive>:/| | it + "<segment>" | segment <- path.segments);
```
I thought we needed to make it into a string first? so `file:///<"<drive>">:/` ?

jurgenvinju replied:
No that's idempotent. Both do string interpolation. The problem is with string interpolation in the path part of a loc; then the encoding feature is broken. So I'm using `+` for the path part.

DavyLandman commented on src/org/rascalmpl/library/lang/paths/Windows.rsc:
```diff
@@ -0,0 +1,113 @@
+    = (|cwdrive:///| | it + "<segment>" | segment <- path.segments);
+
+@synopsis{Relative to the current working directory on the current drive.}
+loc mapPathToLoc((WindowsPath) `<WindowsFilePath path>`) 
```
all these 4 functions are the same, with the exception of the root uri, can we refactor that? so that it becomes something like:

```
loc mapPathToLoc((WindowsPath) `<Drive drive>:<WindowsFilePath path>`) = build(|file:///<drive>:.|, path);
loc mapPathToLoc((WindowsPath) `<WindowsFilePath path>`) = build(|cwd:///|, path);
...
```
?

jurgenvinju replied:
Definitely possible and would make the code less wide. I'll have a look.

DavyLandman commented on src/org/rascalmpl/library/lang/paths/Windows.rsc:
```diff
@@ -0,0 +1,113 @@
+
+private bool IS_WINDOWS = /win/i := getSystemProperty("os.name");
+
+test bool uncSharePath()
```
we need to also have a test case for `\\?\..` and `\\.\`.

jurgenvinju replied:
yes! do you have more ideas for tests?

jurgenvinju replied:
We are not resolving the `..` and `.` btw. Those will end up in the URI. See another newer issue report for adding support for path normalization acros the board.

DavyLandman commented on src/org/rascalmpl/library/lang/paths/Windows.rsc:
```diff
@@ -0,0 +1,113 @@
+    = parseWindowsPath("D:\\Program Files\\Rascal")
+    == |file:///D:/Program%20Files/Rascal|;
+
+test bool uncNetworkShareOk() {
```
if the tests are in this file, are they even tested by our regular JUnitRunners?

DavyLandman commented on src/org/rascalmpl/uri/file/UNCResolver.java:
```diff
@@ -0,0 +1,39 @@
+        if (uri.hasAuthority()) {
+            // downstream methods will use `new File` and `new FileInputStream`
+            // which are able to parse UNC's on Windows.
+			return "\\\\" + uri.getAuthority() + "\\" + uri.getPath();
```
I think we should also flip the slashes in the `getPath` result.

jurgenvinju replied:
Why? Those slashes are supported as-is; it will just slow down the processing.

DavyLandman reviewed (commented):
Looks good, I only have some very minor questions.

jurgenvinju commented:
@DavyLandman this test fails on Windows:

```
test bool uncDOSDrive() {
    loc l = parseWindowsPath("\\\\?\\C$\\");
    
    if (IS_WINDOWS) {
        assert exists(l);
    }

    return l == |unc://%3F/C$|;
}
```

It's probably a small error in the implementation of the UNC scheme; not decoding the `%3F` perhaps? if you have some time later could you run this with a debugger  on a Windows machine? The encoding is required because `?` is not a valid authority character.

DavyLandman commented:
> @DavyLandman this test fails on Windows:
> 
> ```
> test bool uncDOSDrive() {
>     loc l = parseWindowsPath("\\\\?\\C$\\");
>     
>     if (IS_WINDOWS) {
>         assert exists(l);
>     }
> 
>     return l == |unc://%3F/C$|;
> }
> ```
> 
> It's probably a small error in the implementation of the UNC scheme; not decoding the `%3F` perhaps? if you have some time later could you run this with a debugger on a Windows machine? The encoding is required because `?` is not a valid authority character.

The problem is that `\\?\` is a local path, not a notwork path, so `\\?\c:\Program Files\` would be valid.

DavyLandman commented:
Some explorations of UNC DOS Device paths:

```
C:\Users\Davy>DIR c:\tmp
 Volume in drive C has no label.
 Volume Serial Number is C404-106D

 Directory of c:\tmp

19/04/2024  09:29    <DIR>          .
20/05/2023  21:33                 0 agent-0xc0000745b0.sock
20/05/2023  21:06                 0 agent-0xc000074fe0.sock
20/05/2023  21:08                 0 agent-0xc000204460.sock
19/04/2024  09:29                 0 agent-0xc00020c2c0.sock
20/05/2023  21:09                 0 agent-0xc0003023d0.sock
08/05/2024  13:54                54 fruits.txt
               6 File(s)             54 bytes
               1 Dir(s)  23.553.859.584 bytes free

C:\Users\Davy>DIR \\?\c:\tmp
 Volume in drive \\?\c: has no label.
 Volume Serial Number is C404-106D

 Directory of \\?\c:\tmp

19/04/2024  09:29    <DIR>          .
20/05/2023  21:33                 0 agent-0xc0000745b0.sock
20/05/2023  21:06                 0 agent-0xc000074fe0.sock
20/05/2023  21:08                 0 agent-0xc000204460.sock
19/04/2024  09:29                 0 agent-0xc00020c2c0.sock
20/05/2023  21:09                 0 agent-0xc0003023d0.sock
08/05/2024  13:54                54 fruits.txt
               6 File(s)             54 bytes
               1 Dir(s)  23.553.642.496 bytes free

C:\Users\Davy>dir \\.\BootPartition\Windows
 Volume in drive \\.\BootPartition has no label.
 Volume Serial Number is C404-106D

 Directory of \\.\BootPartition\Windows

11/04/2024  03:57    <DIR>          .
14/11/2022  15:34    <DIR>          addins
27/08/2023  23:48    <DIR>          appcompat
10/05/2024  09:29    <DIR>          apppatch
....
```

More examples here: https://medium.com/walmartglobaltech/dos-file-path-magic-tricks-5eda7a7a85fa

```
\\.\c:\Windows\System32\notepad.exe
\\.\BootPartition\Windows\System32\notepad.exe
\\.\HarddiskVolume4\Windows\System32\notepad.exe
\\.\Volume{b289d78f-d491-4f6b-a437-5c51a88fa48f}\Windows\System32\notepad.exe
\\.\Global\BootPartition\Windows\System32\notepad.exe
\\.\GLOBALROOT\Device\HarddiskVolume4\Windows\System32\notepad.exe
\\.\UNC\LOCALHOST\c$\Windows\System32\notepad.exe
\\?\c:\Windows\System32\notepad.exe
```



```

jurgenvinju commented:
I've studied the UNC specs and the examples. The normalization is not good here, that's for another feature of the ISourceLocationResolver interface. 

The reserved file names are not part of the syntax, and different names are reserved for different versions of the OS. What will happen with examples like `COM1` and `LPT1` is that at IO-time, the same exception will be thrown as if the user had typed `new File("\\\\?\LPT1") themselves. A UNC for a communication port in DOS is BTW a valid UNC that can be used for other purposes than file IO (like raw interfacing).

jurgenvinju commented:
I need a little help on the UNC DOS test however. That I can not debug without a running system.

DavyLandman commented:
At the very least the grammar has to be updated, as `:` is a legal char in a DOSDevicePath.

But I'm open to having a pair programming session.

jurgenvinju commented:
Strange; in my documention people always write `$` insteaf of `:` because `:` is not allowed. I'll dig into this again.

DavyLandman commented:
> Strange; in my documention people always write `$` insteaf of `:` because `:` is not allowed. I'll dig into this again.

```
\\localhost\c$
```

is a UNC Network path, pointing towards to a network share named `C$`. This is a default networkshare that windows makes (and maps onto the root c drive. but it could also have been `\\localhost\share-for-jurgen`. The `$` ones are non-user defined shares.

```
\\(.|?)\...
```

Are UNC DOS Device paths, offer access to all the local devices. so `\\.\c:\` is the c drive, but it can also be accessed as `\\.\BootPartition`. Now since the network UNC is also a network device, you can acces that as well via `\\.\UNC\<original network path>`.

jurgenvinju commented:
Right. Thanks. I think I've fixed this for now. Would be great if Microsoft posts RFC's for this kind of stuff. Now I combine reading source code and reading documentation :-) With your help it is becoming a lot better.

I removed the normalization tests. All filesystem notation normalization is going to be deferred to another library function, because this is a complex matter. So for now `c:\` and `C:\` will produce different `loc` values pointing to the same drive.

jurgenvinju commented:
Unfortunately `\\.\BootPartition` is not available on the test machine of github actions. otherwise that would be a nice test as well. The Admin share is there however, so we can make tests for that.

DavyLandman commented on src/org/rascalmpl/library/lang/paths/Unix.rsc:
```diff
@@ -0,0 +1,108 @@
+
+test bool userPath()
+    = parseUnixPath("~root/.bash_profile")
+    == |home:///../root/.bash_profile|;
```
shouldn't this map to `home://root/` ?

jurgenvinju replied:
I tried that but we can not assure cross-platform access to the user directories of a system. It's simply not a portable feature of the Java run-time. So I chose here to transparantly encode the assumption that user directories are siblings. The `home:///` root is resolved using the "user.dir" property that _is_ guaranteed to be portable for the current user. That's the best we can do, or throw out the entire feature.. that has crossed my mind as well.

DavyLandman commented on src/org/rascalmpl/library/lang/paths/Windows.rsc:
```diff
@@ -0,0 +1,206 @@
+
+    
+    if (IS_WINDOWS) {
+        assert exists(l);
```
why would we assume that `c:\test\foo.txt` is a file that exists?

jurgenvinju replied:
that makes no sense indeed. copy/paste error.

jurgenvinju replied:
the C:\ should exist however, but I have not found the reason why that fails yet. Currently trying to install Windows in VirtualBox.

DavyLandman commented:
LGTM, thanks for all the effort of getting it right on Windows.
